src/theory/strings/theory_strings_preprocess.cpp

```diff
@@ -174,106 +174,88 @@ Node StringsPreprocess::simplify( Node t, std::vector< Node > &new_nodes ) {
-    svec.push_back(cc21);
-    svec.push_back(cc3);svec.push_back(cc4);svec.push_back(cc5);
-    Node conc = NodeManager::currentNM()->mkNode(kind::AND, svec);
-    conc = NodeManager::currentNM()->mkNode( kind::IMPLIES, g1, conc );
-    conc = NodeManager::currentNM()->mkNode( kind::FORALL, b1v, conc );
-    conc = NodeManager::currentNM()->mkNode( kind::IMPLIES, nonneg, conc );
-    new_nodes.push_back( conc );
-
-    /*conc = Rewriter::rewrite(NodeManager::currentNM()->mkNode(kind::IMPLIES,
-            NodeManager::currentNM()->mkNode(kind::LT, t[0], d_zero),
-            t.eqNode(NodeManager::currentNM()->mkNode(kind::STRING_CONCAT,
-              NodeManager::currentNM()->mkConst(::CVC4::String("-")), pret))));
-    new_nodes.push_back( conc );*/
-    retNode = pret;
+    argTypes.push_back(nm->integerType());
+    Node u = nm->mkSkolem("U", nm->mkFunctionType(argTypes, nm->integerType()));
+    Node us =
+        nm->mkSkolem("Us", nm->mkFunctionType(argTypes, nm->stringType()));
+    Node ud =
+        nm->mkSkolem("Ud", nm->mkFunctionType(argTypes, nm->stringType()));
+
+    Node lem = n.eqNode(nm->mkNode(APPLY_UF, u, leni));
+    conc.push_back(lem);
+
+    lem = d_zero.eqNode(nm->mkNode(APPLY_UF, u, d_zero));
+    conc.push_back(lem);
+
+    lem = d_empty_str.eqNode(nm->mkNode(APPLY_UF, us, leni));
+    conc.push_back(lem);
+
+    lem = itost.eqNode(nm->mkNode(APPLY_UF, us, d_zero));
+    conc.push_back(lem);
+
+    Node x = nm->mkBoundVar(nm->integerType());
+    Node xbv = nm->mkNode(BOUND_VAR_LIST, x);
+    Node g =
+        nm->mkNode(AND, nm->mkNode(GEQ, x, d_zero), nm->mkNode(LT, x, leni));
+    Node udx = nm->mkNode(APPLY_UF, ud, x);
+    Node ux = nm->mkNode(APPLY_UF, u, x);
+    Node ux1 = nm->mkNode(APPLY_UF, u, nm->mkNode(PLUS, x, d_one));
+    Node c = nm->mkNode(
+        MINUS, nm->mkNode(STRING_CODE, udx), nm->mkConst(Rational(48)));
+    Node usx = nm->mkNode(APPLY_UF, us, x);
+    Node usx1 = nm->mkNode(APPLY_UF, us, nm->mkNode(PLUS, x, d_one));
+
+    Node ten = nm->mkConst(Rational(10));
+    Node eqs = usx.eqNode(nm->mkNode(STRING_CONCAT, udx, usx1));
+    Node eq = ux1.eqNode(nm->mkNode(PLUS, c, nm->mkNode(MULT, ten, ux)));
+    Node cb = nm->mkNode(
+        AND,
+        nm->mkNode(GEQ, c, nm->mkNode(ITE, x.eqNode(d_zero), d_one, d_zero)),
+        nm->mkNode(LT, c, ten));
+
+    lem = nm->mkNode(OR, g.negate(), nm->mkNode(AND, eqs, eq, cb));
+    lem = nm->mkNode(FORALL, xbv, lem);
+    conc.push_back(lem);
+
+    Node nonneg = nm->mkNode(GEQ, n, d_zero);
+
+    lem = nm->mkNode(
+        ITE, nonneg, nm->mkNode(AND, conc), itost.eqNode(d_empty_str));
+    new_nodes.push_back(lem);
+    // assert:
+    // IF n>=0
+    // THEN:
+    //   n = U( len( itost ) ) ^ U( 0 ) = 0 ^
+    //   "" = Us( len( itost ) ) ^ itost = Us( 0 ) ^
+    //   forall x. (x>=0 ^ x < str.len(itost)) =>
+    //     Us( x ) = Ud( x ) ++ Us( x+1 ) ^
+    //     U( x+1 ) = (str.code( Ud( x ) )-48) + 10*U( x ) ^
```

Comment:
Question about the encoding: wouldn't it be more efficient if both `U` and `Us` were of the form `U(x + 1) = ....` and `Us(x + 1) = ...`?

Reply:
I don't think this matters really. The expectation is that this quantified formula, since it is bounded, should be instantiated exhaustively, meaning the constraints on `Us( 0 ) ... Us( len( itost ) )` and likewise `U( 0 )... U( len( itost ) )` are added eagerly. It just so happens that it is the most natural to express definitions Us "backwards" and U "forwards".